whats-that-typeclass-functor/article.md

```diff
@@ -0,0 +1,507 @@
+
+Hence, it’s required to apply it once, like we do with `Either a` or `((,) a)`. In the `(->)` case, it means providing the function argument’s type – `(->) a` or `a ->` (the latter is not valid syntax in Haskell, though).
+
+For example, functions with the following type signatures have Functor instances:
+
+```haskell
+Int -> a
+String -> a
+(Char, Int) -> a
+[Int] -> a
+```
+
+So we've found out what functions can be functors and what data they match. The next question is – what does it mean to `fmap` a function? Intuitively, it’s about changing the function, i.e., the action it performs. But what is the fixed context here?
+
+Let’s construct the type definition of `fmap` for a one-argument function. For an arbitrary Functor `f`, `fmap` is:
+
+```haskell
+fmap :: Functor f => (a -> b) -> f a -> f b
+```
+
+The type of our data is `(->) c`. Therefore, we need to replace `f t` by `(->) c t` or `c -> t`. Consequently, the type signature is:
+
+```haskell
+fmap :: (a -> b) -> (c -> a) -> (c -> b)
+```
+
+Now we can see that the fixed context corresponds to the type of function’s argument `c`. And the value being modified is the function’s return type – we go from `a` to `b` with the help of the `a -> b` function.
+
+All in all, `fmap` for functions transforms what the function returns while keeping the input unchanged! Does that look familiar to you?
+
+<details>
+<summary>Which Haskell function or operator can produce a <code>c -> b</code> function from <code>a -> b</code> and <code>c -> a</code>?</summary>
+
+Function composition:
+
+```haskell
+> :t (.)
+(.) :: (a -> b) -> (c -> a) -> c -> b
+```
+
+Note that
+
+`(a -> b) -> (c -> a) -> c -> b`
+
+is exactly the same as
+
+`(a -> b) -> (c -> a) -> (c -> b)`
+
+since `->` in Haskell is right-associative.
+
+<hr>
+</details>
+
+Cool! `fmap` for Functor is just the composition of functions.
+
+Here’s the `((->) a)` instance implementation:
+
+```haskell
+instance Functor ((->) a) where
+  fmap = (.)
+```
+
+In conclusion, let’s make sure `fmap` is `(.)` for one-argument function `a -> b` with the example below:
+
+```haskell
+-- fmap :: (Int -> Int) -> (Int -> Int) -> Int -> Int
+> fmap (+1) (*2) 2
+5
+> (+1) . (*2) $ 2
+5
+```
+
+It’s uncommon to consider a function a Functor in Haskell. In fact, it’s more of a fancy case. However, you’ll not be puzzled when you come across this magical and strange usage of `fmap` in the future!
+
+## Why is Functor useful?
+
+In general, Functor is not an extraordinary typeclass. Its methods are easy to grasp and to implement.
+
+Nevertheless, a Haskell project can hardly do without a couple of Functor instances. It enables one to do transformations on the wrapped type without knowing anything about the wrapper, and that's very beneficial.
+
+Moreover, Functor is a solid basis and the predecessor of the Applicative typeclass, which further leads to monads. Hence, it brings you one step closer to the dream of many – understanding monads.
```

Comment:
> it brings you one step closer to the dream of many – understanding monads.

I think we should not feed into this idea that monad is such a mystical and hard thing. We can write something like "which further leads to monads, which is a widely used Typeclass in Haskell and understanding it gives you considerably greater command of the language"